File: app/src/main/java/hibernate/v2/testyourandroid/ui/fragment/AppInfoPermissionFragment.java
package hibernate.v2.testyourandroid.ui.fragment;

import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.PermissionGroupInfo;
import android.content.pm.PermissionInfo;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.blankj.utilcode.util.StringUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import hibernate.v2.testyourandroid.C;
import hibernate.v2.testyourandroid.R;
import hibernate.v2.testyourandroid.model.AppItem;
import hibernate.v2.testyourandroid.model.AppPermissionItem;
import hibernate.v2.testyourandroid.model.InfoItem;
import hibernate.v2.testyourandroid.ui.adapter.InfoItemAdapter;

public class AppInfoPermissionFragment extends BaseFragment {
	private final String NO_GROUP = "Ungrouped Permissions";

	@BindView(R.id.rvlist)
	RecyclerView recyclerView;

	private AppItem appItem;

	private HashMap<String, ArrayList<AppPermissionItem>> map = new HashMap<>();

	public static AppInfoPermissionFragment newInstance(AppItem appItem) {
		AppInfoPermissionFragment fragment = new AppInfoPermissionFragment();
		Bundle args = new Bundle();
		args.putParcelable("APP", appItem);
		fragment.setArguments(args);
		return fragment;
	}

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		// Inflate the layout for this fragment
		View rootView = inflater.inflate(R.layout.fragment_info_listview, container, false);
		ButterKnife.bind(this, rootView);
		return rootView;
	}

	@Override
	public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		recyclerView.setLayoutManager(
				new LinearLayoutManager(mContext,
						RecyclerView.VERTICAL, false)
		);
		init();
	}

	private void init() {
		Bundle bundle = getArguments();
		if (bundle != null) {
			appItem = bundle.getParcelable("APP");

			List<InfoItem> list = new ArrayList<>();

			try {
				PackageManager packageManager = mContext.getPackageManager();
				PackageInfo packageInfo = packageManager.getPackageInfo(appItem.getPackageName(), PackageManager.GET_PERMISSIONS);

				/* Get Permissions */
				String[] requestedPermissions = packageInfo.requestedPermissions;

				if (requestedPermissions != null) {
					for (String requestedPermission : requestedPermissions) {
						String permissionGroupLabel = NO_GROUP;
						String permissionLabel = "";
						String permissionDescription = "";

						try {
							PermissionInfo permissionInfo = packageManager.getPermissionInfo(requestedPermission, 0);

							try {
								PermissionGroupInfo permissionGroupInfo = packageManager.getPermissionGroupInfo(permissionInfo.group, 0);
								permissionGroupLabel = permissionGroupInfo.loadLabel(packageManager).toString();
							} catch (PackageManager.NameNotFoundException | NullPointerException ignored) {
							}
							try {
								permissionLabel = permissionInfo.loadLabel(packageManager).toString();
							} catch (NullPointerException ignored) {
							}
							try {
								permissionDescription = permissionInfo.loadDescription(packageManager).toString();
							} catch (NullPointerException ignored) {
							}


						} catch (Exception e) {
							permissionLabel = requestedPermission;
						}

						permissionLabel = StringUtils.upperFirstLetter(permissionLabel);

						AppPermissionItem appPermissionItem = new AppPermissionItem(permissionGroupLabel, permissionLabel, permissionDescription);

						if (!map.containsKey(permissionGroupLabel)) {
							map.put(permissionGroupLabel, new ArrayList<AppPermissionItem>());
						}

						ArrayList<AppPermissionItem> arrayList = map.get(permissionGroupLabel);
						arrayList.add(appPermissionItem);
						map.put(permissionGroupLabel, arrayList);
					}
				}
				ArrayList<String> sortedKeys = new ArrayList<>(map.keySet());
				Collections.sort(sortedKeys);

				sortedKeys.remove(NO_GROUP);
				sortedKeys.add(NO_GROUP);

				for (Object key : sortedKeys) {
					String permissionGroupLabel = (String) key;
					ArrayList<AppPermissionItem> value = map.get(permissionGroupLabel);

					Collections.sort(value, new Comparator<AppPermissionItem>() {
						@Override
						public int compare(AppPermissionItem item1, AppPermissionItem item2) {
							return item1.getPermissionLabel().compareTo(item2.getPermissionLabel());
						}
					});

					StringBuilder permissionLabel = new StringBuilder();

					for (AppPermissionItem s : value) {
						permissionLabel.append(s.getPermissionLabel()).append("\n");
					}

					list.add(new InfoItem(permissionGroupLabel, permissionLabel.toString().trim()));
				}

			} catch (Exception e) {
				list.add(new InfoItem("Fail to fetch the permissions", "Error: -1034"));
			}

			InfoItemAdapter adapter = new InfoItemAdapter(list);

			recyclerView.setAdapter(adapter);
		} else {
			C.notAppFound(mContext);
		}
	}
}

File: app/src/main/java/hibernate/v2/testyourandroid/ui/appinfo/AppInfoPermissionFragment.kt
package hibernate.v2.testyourandroid.ui.appinfo

import android.content.pm.PackageManager
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.blankj.utilcode.util.StringUtils
import hibernate.v2.testyourandroid.databinding.FragmentInfoListviewBinding
import hibernate.v2.testyourandroid.model.AppItem
import hibernate.v2.testyourandroid.model.AppPermissionItem
import hibernate.v2.testyourandroid.model.InfoItem
import hibernate.v2.testyourandroid.ui.appinfo.AppInfoFragment.Companion.ARG_APP
import hibernate.v2.testyourandroid.ui.base.BaseFragment
import hibernate.v2.testyourandroid.ui.base.InfoItemAdapter
import hibernate.v2.testyourandroid.util.Utils.notAppFound
import java.util.ArrayList
import java.util.Comparator
import java.util.HashMap

class AppInfoPermissionFragment : BaseFragment<FragmentInfoListviewBinding>() {

    override fun getViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): FragmentInfoListviewBinding =
        FragmentInfoListviewBinding.inflate(inflater, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val map = HashMap<String, ArrayList<AppPermissionItem>>()

        arguments?.getParcelable<AppItem>(ARG_APP)?.packageName?.let { packageName ->
            context?.let { context ->
                val list: MutableList<InfoItem> = ArrayList()
                try {
                    val packageManager = context.packageManager
                    val packageInfo =
                        packageManager.getPackageInfo(packageName, PackageManager.GET_PERMISSIONS)
                    /* Get Permissions */
                    val requestedPermissions = packageInfo.requestedPermissions
                    val noGroupLabel = "Ungrouped Permissions"
                    if (requestedPermissions != null) {
                        for (requestedPermission in requestedPermissions) {
                            var permissionGroupLabel = noGroupLabel
                            var permissionLabel: String
                            try {
                                val permissionInfo =
                                    packageManager.getPermissionInfo(requestedPermission, 0)

                                try {
                                    val permissionGroupInfo = packageManager.getPermissionGroupInfo(
                                        permissionInfo.group ?: "", 0
                                    )
                                    permissionGroupLabel =
                                        permissionGroupInfo.loadLabel(packageManager).toString()
                                } catch (ignored: PackageManager.NameNotFoundException) {
                                } catch (ignored: NullPointerException) {
                                }
                                permissionLabel =
                                    permissionInfo.loadLabel(packageManager).toString()
                            } catch (e: Exception) {
                                permissionLabel = requestedPermission
                            }
                            permissionLabel = StringUtils.upperFirstLetter(permissionLabel)
                            val arrayList: ArrayList<AppPermissionItem> =
                                map[permissionGroupLabel] ?: arrayListOf()
                            arrayList.add(AppPermissionItem(permissionLabel))
                            map[permissionGroupLabel] = arrayList
                        }
                    }
                    val sortedKeys = ArrayList(map.keys)
                    sortedKeys.sort()
                    sortedKeys.remove(noGroupLabel)
                    sortedKeys.add(noGroupLabel)
                    for (key in sortedKeys) {
                        val permissionGroupLabel = key as String
                        val value = map[permissionGroupLabel]
                        value?.let {
                            value.sortWith(Comparator { item1, item2 ->
                                item1.permissionLabel.compareTo(
                                    item2.permissionLabel
                                )
                            })
                            val permissionLabel = StringBuilder()
                            for (s in value) {
                                permissionLabel.append(s.permissionLabel).append("\n")
                            }
                            list.add(
                                InfoItem(
                                    permissionGroupLabel,
                                    permissionLabel.toString().trim { it <= ' ' })
                            )
                        }
                    }
                } catch (e: Exception) {
                    list.add(InfoItem("Fail to fetch the permissions", "Error: -1034"))
                }
                viewBinding!!.rvlist.adapter = InfoItemAdapter(list)
            }
        } ?: run {
            notAppFound(activity)
        }
    }

    companion object {
        fun newInstance(appItem: AppItem?): AppInfoPermissionFragment {
            return AppInfoPermissionFragment().apply {
                arguments = Bundle().apply { putParcelable(ARG_APP, appItem) }
            }
        }
    }
}